File: courier/app/github.py
import base64
from typing import List

import requests

GITHUB_API = "https://api.github.com"


def _headers(token: str) -> dict:
    return {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }


def _repo_base(owner: str, repo: str) -> str:
    return f"{GITHUB_API}/repos/{owner}/{repo}"


def list_user_repos(token: str) -> List[dict]:
    resp = requests.get(
        f"{GITHUB_API}/user/repos",
        headers=_headers(token),
        params={"sort": "updated", "per_page": 100},
        timeout=20,
    )
    resp.raise_for_status()
    return [
        {
            "full_name": r["full_name"],
            "name": r["name"],
            "owner": r["owner"]["login"],
            "private": r["private"],
            "default_branch": r.get("default_branch", "main"),
        }
        for r in resp.json()
    ]


def list_markdown_files(token: str, owner: str, repo: str, branch: str = "main") -> List[str]:
    url = f"{_repo_base(owner, repo)}/git/trees/{branch}"
    resp = requests.get(url, headers=_headers(token), params={"recursive": "1"}, timeout=20)
    resp.raise_for_status()
    data = resp.json()
    files = [
        item["path"]
        for item in data.get("tree", [])
        if item.get("type") == "blob" and item.get("path", "").endswith(".md")
    ]
    return sorted(files)


def get_file_content(path: str, token: str, owner: str, repo: str, branch: str = "main") -> str:
    url = f"{_repo_base(owner, repo)}/contents/{path}"
    resp = requests.get(url, headers=_headers(token), params={"ref": branch}, timeout=20)
    resp.raise_for_status()
    data = resp.json()
    if data.get("encoding") != "base64":
        raise RuntimeError("Unexpected content encoding from GitHub")
    raw = base64.b64decode(data.get("content", ""))
    return raw.decode("utf-8")


def _get_file_sha(path: str, token: str, owner: str, repo: str, branch: str = "main") -> str:
    url = f"{_repo_base(owner, repo)}/contents/{path}"
    resp = requests.get(url, headers=_headers(token), params={"ref": branch}, timeout=20)
    resp.raise_for_status()
    sha = resp.json().get("sha")
    if not sha:
        raise RuntimeError("Unable to fetch file sha")
    return sha


def commit_file(
    path: str,
    content: str,
    message: str,
    token: str,
    owner: str,
    repo: str,
    branch: str = "main",
) -> dict:
    url = f"{_repo_base(owner, repo)}/contents/{path}"
    payload = {
        "message": message or f"Update {path}",
        "content": base64.b64encode(content.encode("utf-8")).decode("utf-8"),
        "sha": _get_file_sha(path, token, owner, repo, branch),
        "branch": branch,
    }
    resp = requests.put(url, headers=_headers(token), json=payload, timeout=20)
    resp.raise_for_status()
    return resp.json()


def create_file(
    path: str,
    content: str,
    message: str,
    token: str,
    owner: str,
    repo: str,
    branch: str = "main",
) -> dict:
    url = f"{_repo_base(owner, repo)}/contents/{path}"
    payload = {
        "message": message or f"Create {path}",
        "content": base64.b64encode(content.encode("utf-8")).decode("utf-8"),
        "branch": branch,
    }
    resp = requests.put(url, headers=_headers(token), json=payload, timeout=20)
    resp.raise_for_status()
    return resp.json()


def delete_file(
    path: str,
    message: str,
    token: str,
    owner: str,
    repo: str,
    branch: str = "main",
) -> dict:
    url = f"{_repo_base(owner, repo)}/contents/{path}"
    payload = {
        "message": message or f"Delete {path}",
        "sha": _get_file_sha(path, token, owner, repo, branch),
        "branch": branch,
    }
    resp = requests.delete(url, headers=_headers(token), json=payload, timeout=20)
    resp.raise_for_status()
    return resp.json()

File: worker/src/github.ts
const GITHUB_API = 'https://api.github.com'

export interface RepoInfo {
  full_name: string
  name: string
  owner: string
  private: boolean
  default_branch: string
}

function headers(token: string): Record<string, string> {
  return {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
    // GitHub API は User-Agent 必須。Lambda では requests が自動付与していたが fetch は付けないため明示する。
    'User-Agent': 'quill-md-editor',
  }
}

function repoBase(owner: string, repo: string): string {
  return `${GITHUB_API}/repos/${owner}/${repo}`
}

async function ghFetch(url: string, token: string, init?: RequestInit): Promise<Response> {
  const res = await fetch(url, { ...init, headers: { ...headers(token), ...(init?.headers ?? {}) } })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`GitHub API ${res.status}: ${body}`)
  }
  return res
}

export async function listUserRepos(token: string): Promise<RepoInfo[]> {
  const res = await ghFetch(`${GITHUB_API}/user/repos?sort=updated&per_page=100`, token)
  const repos = (await res.json()) as any[]
  return repos.map((r) => ({
    full_name: r.full_name,
    name: r.name,
    owner: r.owner.login,
    private: r.private,
    default_branch: r.default_branch ?? 'main',
  }))
}

export async function listMarkdownFiles(
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<string[]> {
  const res = await ghFetch(`${repoBase(owner, repo)}/git/trees/${branch}?recursive=1`, token)
  const data = (await res.json()) as any
  const tree = (data.tree ?? []) as any[]
  return tree
    .filter((item) => item.type === 'blob' && typeof item.path === 'string' && item.path.endsWith('.md'))
    .map((item) => item.path as string)
    .sort()
}

export async function getFileContent(
  path: string,
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<string> {
  const url = `${repoBase(owner, repo)}/contents/${encodeURIComponent(path)}?ref=${encodeURIComponent(branch)}`
  const res = await ghFetch(url, token)
  const data = (await res.json()) as any
  if (data.encoding !== 'base64') {
    throw new Error('Unexpected content encoding from GitHub')
  }
  // GitHub は改行入りの base64 を返すため、そのまま atob できるよう改行を除去してからデコードする。
  const bin = atob(String(data.content ?? '').replace(/\n/g, ''))
  const bytes = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i)
  return new TextDecoder('utf-8').decode(bytes)
}

async function getFileSha(
  path: string,
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<string> {
  const url = `${repoBase(owner, repo)}/contents/${encodeURIComponent(path)}?ref=${encodeURIComponent(branch)}`
  const res = await ghFetch(url, token)
  const data = (await res.json()) as any
  if (!data.sha) {
    throw new Error('Unable to fetch file sha')
  }
  return data.sha as string
}

// UTF-8 文字列を base64 化する。btoa は Latin-1 前提なので、一度 UTF-8 バイト列へ変換してから通す。
function toBase64(content: string): string {
  const bytes = new TextEncoder().encode(content)
  let bin = ''
  for (const b of bytes) bin += String.fromCharCode(b)
  return btoa(bin)
}

export async function commitFile(
  path: string,
  content: string,
  message: string,
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<any> {
  const url = `${repoBase(owner, repo)}/contents/${encodeURIComponent(path)}`
  const res = await ghFetch(url, token, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      message: message || `Update ${path}`,
      content: toBase64(content),
      sha: await getFileSha(path, token, owner, repo, branch),
      branch,
    }),
  })
  return res.json()
}

export async function createFile(
  path: string,
  content: string,
  message: string,
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<any> {
  const url = `${repoBase(owner, repo)}/contents/${encodeURIComponent(path)}`
  const res = await ghFetch(url, token, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      message: message || `Create ${path}`,
      content: toBase64(content),
      branch,
    }),
  })
  return res.json()
}

export async function deleteFile(
  path: string,
  message: string,
  token: string,
  owner: string,
  repo: string,
  branch = 'main',
): Promise<any> {
  const url = `${repoBase(owner, repo)}/contents/${encodeURIComponent(path)}`
  const res = await ghFetch(url, token, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      message: message || `Delete ${path}`,
      sha: await getFileSha(path, token, owner, repo, branch),
      branch,
    }),
  })
  return res.json()
}
